c/src/atom.rs

```diff
@@ -644,49 +649,141 @@ pub unsafe extern "C" fn atom_eq(atoma: *const atom_ref_t, atomb: *const atom_re
+        unsafe{ core::slice::from_raw_parts(self.ptr.cast(), self.len) }
+    }
+}
 
 #[no_mangle]
-pub extern "C" fn vec_atom_new() -> *mut vec_atom_t {
-    vec_atom_into_ptr(Vec::new())
+pub extern "C" fn atom_vec_new() -> atom_vec_t {
+    atom_vec_t::new()
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn vec_atom_free(vec: *mut vec_atom_t) {
-    drop(Box::from_raw(vec));
+pub unsafe extern "C" fn atom_vec_from_array(atoms: *mut atom_t, size: usize) -> atom_vec_t {
+    let c_arr = std::slice::from_raw_parts_mut(atoms, size);
+    let atoms: Vec<Atom> = c_arr.into_iter().map(|atom| {
+        core::mem::replace(atom, atom_t::null()).into_inner()
+    }).collect();
+    atoms.into()
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn vec_atom_size(vec: *mut vec_atom_t) -> usize {
-    (*vec).0.len()
+pub unsafe extern "C" fn atom_vec_free(vec: atom_vec_t) {
+    drop(vec);
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn vec_atom_pop(vec: *mut vec_atom_t) -> atom_t {
-    (&mut *vec).0.pop().expect("Vector is empty").into()
+pub unsafe extern "C" fn atom_vec_len(vec: *const atom_vec_t) -> usize {
+    (*vec).len
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn vec_atom_push(vec: *mut vec_atom_t, atom: atom_t) {
-    (*vec).0.push(atom.into_inner());
+pub unsafe extern "C" fn atom_vec_pop(vec: *mut atom_vec_t) -> atom_t {
+    let vec_contents = core::mem::replace(&mut *vec, core::mem::zeroed());
+    if !vec_contents.owned {
+        panic!("Error! Attempt to modify read-only atom_vec_t");
+    }
+    let mut rust_vec: Vec<Atom> = vec_contents.into();
+    let result_atom: atom_t = rust_vec.pop().into();
+    *vec = rust_vec.into();
+    result_atom
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn vec_atom_len(vec: *const vec_atom_t) -> usize {
-    (*vec).0.len()
+pub unsafe extern "C" fn atom_vec_push(vec: *mut atom_vec_t, atom: atom_t) {
+    let vec_contents = core::mem::replace(&mut *vec, core::mem::zeroed());
+    if !vec_contents.owned {
+        panic!("Error! Attempt to modify read-only atom_vec_t");
+    }
+    let mut rust_vec: Vec<Atom> = vec_contents.into();
+    rust_vec.push(atom.into_inner());
+    *vec = rust_vec.into();
 }
 
-/// WARNING: The atom returned from this function remains owned by the vec_atom_t.  It must NOT be
-/// accessed after the vec_atom_get has been modified or freed
+/// WARNING: The atom returned from this function remains owned by the atom_vec_t.  It must NOT be
+/// accessed after the atom_vec_get has been modified or freed
 #[no_mangle]
-pub unsafe extern "C" fn vec_atom_get(vec: *const vec_atom_t, idx: usize) -> atom_ref_t {
-    let atom = &(*vec).0[idx];
+pub unsafe extern "C" fn atom_vec_get(vec: *const atom_vec_t, idx: usize) -> atom_ref_t {
+    let vec = &*vec;
+    let atom: &Atom = &vec.as_slice()[idx];
     atom.into()
 }
 
-pub type atom_array_t = array_t<atom_ref_t>;
-pub type c_atoms_callback_t = lambda_t<atom_array_t>;
+pub type c_atoms_callback_t = lambda_t<*const atom_vec_t>;
```

Comment:
I would suggest to rename it into `c_atom_vec_callback_t`.

Reply:
Name changed in 2059746